[Java] Add serialization support for  java.util.ArrayList$SubList

When performing serialization with any Java object contains a List field produced by Java subList method of List, an application will get this error "java.lang.UnsupportedOperationException: Class class java.util.ArrayList$SubList doesn't support serialization."

It'd be nice to provide support for this particular serialization without resulting to create customer serialization for this standard JDK feature.

This issue can be recrreated through the following test class.
```java
import io.fury.Fury;
import io.fury.ThreadSafeFury;
import org.junit.jupiter.api.Test;

import java.util.ArrayList;
import java.util.List;

import static org.junit.jupiter.api.Assertions.assertEquals;

public class FuryTest
{
  private static final ThreadSafeFury fury = Fury.builder()
    .requireClassRegistration(false)
    .buildThreadSafeFury();

  @Test
  public void serializeAndDeserializeSubList()
  {
    List<Integer> numbers = new ArrayList<>(List.of(1, 2, 3, 4, 5));

    List<List<Integer>> partitions =
      List.of(numbers.subList(0,1), numbers.subList(1,2), numbers.subList(2, numbers.size()));

    byte[] serialized = fury.serialize(partitions);
    assertEquals(partitions, fury.deserialize(serialized));
  }
}
```

Here is the part of the stack trace:
```java
java.lang.UnsupportedOperationException: Class class java.util.ArrayList$SubList doesn't support serialization.
	at io.fury.resolver.ClassResolver.getSerializerClass(ClassResolver.java:838)
	at io.fury.resolver.ClassResolver.getSerializerClass(ClassResolver.java:791)
	at io.fury.resolver.ClassResolver.createSerializer(ClassResolver.java:1182)
	at io.fury.resolver.ClassResolver.getClassInfo(ClassResolver.java:1081)
	at io.fury.serializer.collection.AbstractCollectionSerializer.writeTypeNullabilityHeader(AbstractCollectionSerializer.java:284)
	at io.fury.serializer.collection.AbstractCollectionSerializer.writeElementsHeader(AbstractCollectionSerializer.java:144)
	at io.fury.serializer.collection.AbstractCollectionSerializer.writeElements(AbstractCollectionSerializer.java:301)
	at io.fury.serializer.collection.AbstractCollectionSerializer.write(AbstractCollectionSerializer.java:296)
	at io.fury.Fury.writeData(Fury.java:551)
	at io.fury.Fury.writeRef(Fury.java:344)
	at io.fury.Fury.write(Fury.java:319)
	at io.fury.Fury.serialize(Fury.java:255)
	at io.fury.Fury.serialize(Fury.java:221)
	at io.fury.ThreadLocalFury.serialize(ThreadLocalFury.java:67)
```

Don't know if this is the right place to ask this question. I just checked out the project and ran ` mvn -T10 clean test`.
There are some unit test failures. Any ideas?

Here are the errors
[ERROR] Failures: 
[ERROR]   CrossLanguageTest.testBuffer:135 expected [true] but found [false]
[ERROR]   CrossLanguageTest.testCrossLanguageReference:433 expected [true] but found [false]
[ERROR]   CrossLanguageTest.testCrossLanguageSerializer:308 expected [true] but found [false]
[ERROR]   CrossLanguageTest.testMurmurHash3:167 expected [true] but found [false]
[ERROR]   CrossLanguageTest.testOutOfBandBuffer:695 expected [true] but found [false]
[ERROR]   CrossLanguageTest.testRegisterSerializer:641 expected [true] but found [false]
[ERROR]   CrossLanguageTest.testSerializeComplexStruct:501->structRoundBack:515 expected [true] but found [false]
[ERROR]   CrossLanguageTest.testSerializeOpaqueObject:562 expected [true] but found [false]
[ERROR]   CrossLanguageTest.testSerializeSimpleStruct:472->structRoundBack:515 expected [true] but found [false]
[INFO]
[ERROR] Tests run: 797, Failures: 9, Errors: 0, Skipped: 3

> CrossLanguageTest.testSerializeSimpleStruct:472->structRoundBack:515 expected [true] but found [false] [INFO] [ERROR] Tests run: 797, Failures: 9, Errors: 0, Skipped: 3

Such cases needs to install `pyfury` first, you can ignore it if you don't develop cross-language features. `pyfury` can be installed by following  https://github.com/alipay/fury/blob/main/docs/guide/DEVELOPMENT.md#building-fury-python . You may need to setup gcc/clang and pyenv for install pyfury too, which may take some time.

--- a/java/fury-core/src/main/java/org/apache/fury/serializer/collection/SubListSerializers.java
+++ b/java/fury-core/src/main/java/org/apache/fury/serializer/collection/SubListSerializers.java
@@ -78,8 +78,8 @@ public class SubListSerializers {
         new Class[] {
           SubListClass, RandomAccessSubListClass, ArrayListSubListClass, ImmutableSubListClass
         }) {
-      if (preserveView && fury.getConfig().getLanguage() == Language.JAVA) {
-        fury.registerSerializer(cls, new SubListViewSerializer(fury, (Class<List>) cls));
+      if (fury.trackingRef() && preserveView && fury.getConfig().getLanguage() == Language.JAVA) {
+        fury.registerSerializer(cls, new SubListViewSerializer(fury, cls));
       } else {
         fury.registerSerializer(cls, new SubListSerializer(fury, (Class<List>) cls));
       }

--- a/java/fury-core/src/test/java/org/apache/fury/serializer/collection/SubListSerializersTest.java
+++ b/java/fury-core/src/test/java/org/apache/fury/serializer/collection/SubListSerializersTest.java
@@ -32,7 +32,7 @@ import org.testng.annotations.Test;
 public class SubListSerializersTest extends FuryTestBase {
   @Test
   public void testSubListViewSerialization() {
-    Fury fury = builder().build();
+    Fury fury = builder().withRefTracking(true).build();
     List<Integer> data = new ArrayList<>();
     Collections.addAll(data, 1, 2, 3, 4, 5, 6, 7);
     int length = fury.serialize(data).length;
@@ -65,8 +65,7 @@ public class SubListSerializersTest extends FuryTestBase {
 
   @Test
   public void testSubListNoViewSerialization() {
-    Fury fury = builder().build();
-    SubListSerializers.registerSerializers(fury, false);
+    Fury fury = builder().withRefTracking(false).build();
     List<Integer> data = new ArrayList<>();
     Collections.addAll(data, 1, 2, 3, 4, 5, 6, 7);
     int length = fury.serialize(data).length;

